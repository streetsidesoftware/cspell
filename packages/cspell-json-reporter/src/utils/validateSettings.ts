import { AssertionError } from 'assert';

import type { CSpellJSONReporterSettings } from '../CSpellJSONReporterSettings';

function assertBooleanOrUndefined(key: string, value: unknown): asserts value is boolean | undefined {
    if (typeof value !== 'boolean' && value !== undefined) {
        throw new AssertionError({
            message: `cspell-json-reporter settings.${key} must be a boolean`,
            actual: typeof value,
            expected: 'boolean',
        });
    }
}

/**
 * Throws an error if passed cspell-json-reporter settings are invalid
 */
export function validateSettings(settings: unknown): asserts settings is CSpellJSONReporterSettings {
    if (!settings || typeof settings !== 'object') {
        throw new AssertionError({
            message: 'cspell-json-reporter settings must be an object',
            actual: typeof settings,
            expected: 'object',
        });
    }

    const { outFile, debug, verbose, progress } = settings as CSpellJSONReporterSettings;

    if (typeof outFile !== 'string') {
        throw new AssertionError({
            message: 'cspell-json-reporter settings.outFile must be a string',
            actual: typeof outFile,
            expected: 'string',
        });
    }

    assertBooleanOrUndefined('verbose', verbose);
    assertBooleanOrUndefined('debug', debug);
    assertBooleanOrUndefined('progress', progress);
}
