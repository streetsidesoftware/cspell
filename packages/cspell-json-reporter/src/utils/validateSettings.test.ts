import { validateSettings } from './validateSettings';

describe('validateSettings', () => {
    test.each`
        settings
        ${{ outFile: 'foobar' }}
        ${{ outFile: 'foobar', verbose: true }}
        ${{ outFile: 'foobar', debug: false }}
        ${{ outFile: 'foobar', progress: true }}
        ${{ outFile: 'foobar', verbose: undefined }}
    `('passes valid settings', ({ settings }) => {
        expect(() => validateSettings(settings)).not.toThrow();
    });

    test.each`
        settings
        ${[]}
        ${false}
        ${true}
        ${'hello'}
        ${{ outFile: {} }}
        ${{ outFile: 1 }}
        ${{ outFile: 'foobar', verbose: 123 }}
        ${{ outFile: 'foobar', debug: [] }}
    `('throws for invalid settings $settings', ({ settings }) => {
        expect(() => validateSettings(settings)).toThrowErrorMatchingSnapshot();
    });
});
