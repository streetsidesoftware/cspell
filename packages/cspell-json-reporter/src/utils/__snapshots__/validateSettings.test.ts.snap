// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validateSettings > throws for invalid settings 'hello' 1`] = `[AssertionError: cspell-json-reporter settings must be an object]`;

exports[`validateSettings > throws for invalid settings [] 1`] = `[AssertionError: cspell-json-reporter settings must be an object]`;

exports[`validateSettings > throws for invalid settings { outFile: 'foobar', debug: [] } 1`] = `[AssertionError: cspell-json-reporter settings.debug must be a boolean]`;

exports[`validateSettings > throws for invalid settings { outFile: 'foobar', verbose: 123 } 1`] = `[AssertionError: cspell-json-reporter settings.verbose must be a boolean]`;

exports[`validateSettings > throws for invalid settings { outFile: {} } 1`] = `[AssertionError: cspell-json-reporter settings.outFile must be a string]`;

exports[`validateSettings > throws for invalid settings { outFile: 1 } 1`] = `[AssertionError: cspell-json-reporter settings.outFile must be a string]`;

exports[`validateSettings > throws for invalid settings false 1`] = `[AssertionError: cspell-json-reporter settings must be an object]`;

exports[`validateSettings > throws for invalid settings true 1`] = `[AssertionError: cspell-json-reporter settings must be an object]`;
